design/design-doc-composition-validating-webhook.md

```diff
@@ -0,0 +1,359 @@
+  we render valid Composed Resources if defined.
+- `writeConnectionSecretsToNamespace` : the name of the namespace the `Secret`
+  containing the connection details should be created into. We could check
+  whether it exists already, but that would be against the behaviour Kubernetes
+  users are used to, e.g. a `Deployment` mounting a `Secret` which does not
+  exist yet is considered valid but remains pending until it’s not created.
+- `publishConnectionDetailsWithStoreConfigRef` : a reference by name to a
+  `StoreConfig`, we could check that it exists, but similarly to the namespace
+  it would go against the behaviour Kubernetes users might be expecting.
+
+### Additional Constraints for the ValidatingWebhook use case
+
+A few constraints that makes validating Compositions from a Validating Webhook
+harder w.r.t. the cases mentioned in the Background:
+
+1. Managed Resources defined as `base` and traversed by `patches` might not yet
+   have been deployed when the Composition is being validated.
+2. Composite Resources specified by `CompositeTypeRef` and traversed by
+   `patches` might not yet have been deployed when the Composition is being
+   validated.
+3. We do not have all the required schemas locally, so we should get them from
+   the API server, if available.
+4. We do not have actual input values supplied from the Composite Resource that
+   are required to render valid and applicable Managed Resources.
+5. Some `alpha` features could require some additional interactions with
+   external systems, e.g. `CompositionFunctions`.
+
+### Solution Proposed
+
+In order to be able to have Compositions properly validated we should cover all
+the `resources` parts, while also taking into consideration all the constraints
+above.
+
+So, I propose to implement a Composition Validating Webhook by validating each
+`resource`'s `patches` and final rendered `Composed Resource`, querying the API
+Server to get the needed schemas.
+
+I suggest we avoid validating `bases`, that would require to implement
+validation ignoring required fields, as those could be set at a later stage by
+patches. Assuming we implement full `patches` type coverage, validating `bases`
+per se can be avoided, as any error should either be caught validating the
+patches or can be attributed to the `base` itself. But this could be considered
+at a later stage.
+
+To validate `patches`,  we will need to implement ad-hoc validation functions,
+because the current patching logic is not taking into consideration schemas, as
+it’s dealing directly with unstructured data, therefore not reporting any
+schema related error. The code implemented for
+[upbound/upjet][upject-validation-code] or
+[crossplane-contrib/crossplane-lint][crossplane-lint-validation-code] could be
+good starting points for this.
+
+To render the Composed Resources, we will need to refactor the Composite
+resources reconciliation loop, extracting the rendering logic to remove any
```

Comment:
> extracting the rendering logic

This could be tricky with the `functions` support since rendering logic would also include running the functions.

Reply:
Indeed, updated the doc to explicitly mention functions.